#1018 Migration: Append_GetRange (#1095)
DiceDB/dice: docs/src/content/docs/commands/GETRANGE.MD

---
title: GETRANGE
description: The `GETRANGE` command in DiceDB is used get a substring of string provided the start and end.
---

> The `GETRANGE` command in DiceDB is used get a substring of string provided the start and end.

Reviewer: Shall we draft it as?
`The GETRANGE command in DiceDB is used to get a substring of a string, provided the start and end indices`

## Syntax

```bash
GETRANGE key start end
```

## Parameters
| Parameter | Description                                                               | Type    | Required |
|-----------|---------------------------------------------------------------------------|---------|----------|

> | `key`     | The name of the key to be set.                                            | String  | Yes      |

Reviewer: Shall this be?
`The name of the key containing the string.`

> | `key`     | The name of the key to be set.                                            | String  | Yes      |

Reviewer 2: ```suggestion
| `key`     | The name of the key containing the string.                                           | String  | Yes      |
```

> | `start`   | The starting index of the required substring                              | Integer | Yes      |

Reviewer: `substring` -> `substring.`

| `end`     | The ending index of the required substring.                               | Integer | Yes      |
## Return values

| Condition                                      | Return Value                                      |
|------------------------------------------------|---------------------------------------------------|

> | if key contains a valid stirng                 | a substring based on the start and end indices    |

Reviewer: `stirng` -> `string`

| if `start` is greater than `end`               | an empty string is returned                       |
| if the `end` exceeds the length of the string present at `key` | the entire string is returned     |
| if the `start` is greater than the length of the string | an empty string is returned.             |

## Behaviour
- If the specified key does not exist, the `GETRANGE` command returns an empty string. 
- If `start` is greater than `end`, the `GETRANGE` command returns an empty string.
- If `start` is not within the length of the string, the `GETRANGE` command returns an empty string.

> - `start` and `end` can be negative which removed `end + 1` characters from the other side of the string.

Reviewer: Shall we change it to?
`Both start and end can be negative, which removes characters from the string, starting from the end + 1 position on the right.`

- If both `start` and `end` are within the string's length, the resulting substring includes all characters from start to end, inclusive.

> ## Errors`

Author: [5ec96be](https://github.com/DiceDB/dice/pull/1095/commits/5ec96beeff2cbe27ae4dc6e9271e94ccc3169c35)

1. `Wrong type of value or key`:

   - Error Message: `(error) ERROR WRONGTYPE Operation against a key holding the wrong kind of value`
   - Occurs when attempting to use the command on a key that contains a non-string value.

2. `Invalid syntax or conflicting options`:

   - Error Message: `(error) ERROR wrong number of arguments for 'GETRANGE' command`
   - If the number of arguments are not exactly equal to 4.

> 3. `Invalid Type given as input for start and end`

Reviewer: `Invalid input type for start and end`

   - Error Message: `(error) ERROR value is not an integer or out of range`
   - If `start` and `end` are not integers.

## Example Usage

### Basic Usage

Assume we have a string stored in `foo`

```bash
127.0.0.1:7379> SET foo apple
OK
```

```bash
127.0.0.1:7379> GETRANGE foo 1 3
"ppl"
```

```bash
127.0.0.1:7379> GETRANGE foo 0 -1
"apple"
```

```bash
127.0.0.1:7379> GETRANGE foo 0 -10
""
```

```bash
127.0.0.1:7379> GETRANGE foo 0 -2
"appl"
```

```bash
127.0.0.1:7379> GETRANGE foo 0 1001
"apple"
```

### Invalid usage

Trying to use `GETRANGE` without giving the value

```bash
127.0.0.1:7379> GETRANGE foo
(error) ERROR wrong number of arguments for 'getrange' command
```

> Trying to use `GETRANGE` on a invalid data type.

Reviewer: `Trying to use `GETRANGE` on an invalid data type.`

> Trying to use `GETRANGE` on an invalid data type..

Reviewer 2: ```suggestion
Trying to use `GETRANGE` on an invalid data type.
```
